File: curriculum/12_distributed_job_scheduler/node-impl/tests/scheduler.test.ts
import { describe, expect, it } from 'vitest';
import {
  Election,
  JobResult,
  JobStatus,
  LockManager,
  Priority,
  Scheduler,
} from '../src/scheduler.js';

describe('distributed job scheduler core', () => {
  it('validates duration intervals and tracks submitted status', () => {
    const scheduler = new Scheduler('node-a', () => 0);
    const job = scheduler.submit({ name: 'digest', interval: '5s', priority: Priority.Normal });
    const minuteJob = scheduler.submit({ name: 'minute', interval: '1m', priority: Priority.Low });

    expect(job.status).toBe(JobStatus.Pending);
    expect(job.intervalMs).toBe(5_000);
    expect(minuteJob.intervalMs).toBe(60_000);
    expect(() => scheduler.submit({ name: 'bad', interval: 'cron * * *', priority: Priority.Low })).toThrow(/invalid interval/i);
    expect(() => scheduler.submit({ name: 'zero', interval: '0s', priority: Priority.Low })).toThrow(/invalid interval/i);
    expect(() => scheduler.submit({ name: 'missing-parent', dependencies: ['missing'] })).toThrow(/dependency missing/i);
  });

  it('elects the highest process id leader with a lease', () => {
    let now = 0;
    const election = new Election(() => now, 2_000);

    expect(election.elect(['pid-2', 'pid-9', 'pid-3'])).toBe('pid-9');
    expect(election.isLeader('pid-9')).toBe(true);
    now += 3_000;
    expect(election.isLeader('pid-9')).toBe(false);
  });

  it('dispatches by priority, due time, and creation order', () => {
    const scheduler = new Scheduler('node-a', () => 0);
    scheduler.becomeLeader(['node-a'], 60_000);
    const low = scheduler.submit({ name: 'low', priority: Priority.Low });
    const highLate = scheduler.submit({ name: 'high-late', priority: Priority.High, runAfterMs: 2_000 });
    const highEarly = scheduler.submit({ name: 'high-early', priority: Priority.High });

    expect(scheduler.dispatchNext('worker-1', 1_000).job.id).toBe(highEarly.id);
    expect(scheduler.dispatchNext('worker-1', 1_000).job.id).toBe(highLate.id);
    expect(scheduler.dispatchNext('worker-1', 1_000).job.id).toBe(low.id);
  });

  it('rejects concurrent locks and stale fencing tokens', () => {
    let now = 0;
    const locks = new LockManager(() => now);
    const lock = locks.acquire('job-1', 'leader-a', 'worker-a', 1_000);

    expect(() => locks.acquire('job-1', 'leader-b', 'worker-b', 1_000)).toThrow(/locked/i);
    expect(() => locks.validate('job-1', lock.fencingToken - 1)).toThrow(/stale/i);
    now += 2_000;
    expect(locks.acquire('job-1', 'leader-b', 'worker-b', 1_000).fencingToken).toBeGreaterThan(lock.fencingToken);
  });

  it('rejects stale leadership and expired lock completion', () => {
    let now = 0;
    const scheduler = new Scheduler('node-a', () => now);
    const job = scheduler.submit({ name: 'lease-sensitive' });

    expect(() => scheduler.dispatchJob(job.id, 'worker-1', 1_000)).toThrow(/not leader/i);

    scheduler.becomeLeader(['node-a'], 5_000);
    const dispatched = scheduler.dispatchJob(job.id, 'worker-1', 1_000);
    now += 2_000;
    expect(() => scheduler.complete(job.id, dispatched.lock.fencingToken, JobResult.Completed)).toThrow(/expired/i);
  });

  it('handles DAG dependencies, retry backoff, and cancellation', () => {
    const scheduler = new Scheduler('node-a', () => 0);
    scheduler.becomeLeader(['node-a'], 60_000);
    const parent = scheduler.submit({ name: 'parent', maxAttempts: 1 });
    const child = scheduler.submit({ name: 'child', priority: Priority.High, dependencies: [parent.id], maxAttempts: 1 });

    const dispatchedParent = scheduler.dispatchNext('worker-1', 1_000);
    expect(dispatchedParent.job.id).toBe(parent.id);
    scheduler.complete(parent.id, dispatchedParent.lock.fencingToken, JobResult.Completed);
    expect(scheduler.dispatchNext('worker-1', 1_000).job.id).toBe(child.id);

    const retry = scheduler.submit({ name: 'retry', maxAttempts: 3, initialBackoffMs: 1_000 });
    const dispatchedRetry = scheduler.dispatchJob(retry.id, 'worker-2', 1_000);
    scheduler.complete(retry.id, dispatchedRetry.lock.fencingToken, JobResult.Failed, 'temporary');
    expect(scheduler.getJob(retry.id).status).toBe(JobStatus.Pending);
    expect(scheduler.getJob(retry.id).dueAtMs).toBe(1_000);

    const permanentFailure = scheduler.submit({ name: 'fail', maxAttempts: 1 });
    const dispatchedFailure = scheduler.dispatchJob(permanentFailure.id, 'worker-3', 1_000);
    scheduler.complete(permanentFailure.id, dispatchedFailure.lock.fencingToken, JobResult.Failed, 'permanent');
    expect(scheduler.getJob(permanentFailure.id).status).toBe(JobStatus.Failed);

    const cancelled = scheduler.submit({ name: 'cancel-me' });
    scheduler.cancel(cancelled.id, 'client request');
    expect(scheduler.getJob(cancelled.id).status).toBe(JobStatus.Cancelled);
    expect(() => scheduler.cancel(cancelled.id, 'again')).toThrow(/terminal/i);
  });

  it('reports leader, queue, running jobs, and lock health', () => {
    const scheduler = new Scheduler('node-a', () => 0);
    scheduler.becomeLeader(['node-a'], 60_000);
    scheduler.submit({ name: 'queued' });
    scheduler.dispatchNext('worker-1', 1_000);

    expect(scheduler.health()).toMatchObject({
      nodeId: 'node-a',
      leaderId: 'node-a',
      runningJobs: 1,
      queueDepth: 0,
    });
  });
});

File: curriculum/12_distributed_job_scheduler/go-impl/internal/scheduler/scheduler_test.go
package scheduler

import (
	"context"
	"testing"
	"time"
)

func TestSubmitValidatesIntervalAndTracksStatus(t *testing.T) {
	s := NewScheduler("node-a", Clock{Now: time.Now})

	job, err := s.Submit(context.Background(), JobRequest{
		Name:        "email-digest",
		Payload:     map[string]string{"tenant": "school"},
		Interval:    "5s",
		Priority:    PriorityNormal,
		MaxAttempts: 2,
	})
	if err != nil {
		t.Fatalf("submit valid job: %v", err)
	}
	if job.Status != StatusPending {
		t.Fatalf("status = %s, want %s", job.Status, StatusPending)
	}
	if job.Interval != 5*time.Second {
		t.Fatalf("interval = %s, want 5s", job.Interval)
	}

	if _, err := s.Submit(context.Background(), JobRequest{Name: "bad", Interval: "cron * * *", Priority: PriorityLow}); err == nil {
		t.Fatal("expected invalid interval error")
	}
}

func TestLeaderElectionUsesHighestProcessIDWithLease(t *testing.T) {
	now := time.Date(2026, 6, 17, 12, 0, 0, 0, time.UTC)
	clock := Clock{Now: func() time.Time { return now }}
	e := NewElection(clock, 2*time.Second)

	if leader := e.Elect([]string{"pid-2", "pid-9", "pid-3"}); leader != "pid-9" {
		t.Fatalf("leader = %s, want pid-9", leader)
	}
	if !e.IsLeader("pid-9") {
		t.Fatal("pid-9 should hold leader lease")
	}
	now = now.Add(3 * time.Second)
	if e.IsLeader("pid-9") {
		t.Fatal("expired leader lease should not be active")
	}
}

func TestDispatchOrdersByPriorityDueTimeAndCreation(t *testing.T) {
	now := fixedTime()
	s := NewScheduler("node-a", Clock{Now: func() time.Time { return now }})
	s.BecomeLeader([]string{"node-a"}, time.Minute)

	low, _ := s.Submit(context.Background(), JobRequest{Name: "low", Priority: PriorityLow})
	highLate, _ := s.Submit(context.Background(), JobRequest{Name: "high-late", Priority: PriorityHigh, RunAfter: 2 * time.Second})
	highEarly, _ := s.Submit(context.Background(), JobRequest{Name: "high-early", Priority: PriorityHigh})

	got := []string{}
	for range 3 {
		job, _, err := s.DispatchNext(context.Background(), "worker-1", time.Second)
		if err != nil {
			t.Fatalf("dispatch: %v", err)
		}
		got = append(got, job.ID)
	}
	want := []string{highEarly.ID, highLate.ID, low.ID}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("dispatch order = %v, want %v", got, want)
		}
	}
}

func TestDistributedLockRejectsConcurrentAndStaleTokens(t *testing.T) {
	now := fixedTime()
	lm := NewLockManager(Clock{Now: func() time.Time { return now }})

	lock, err := lm.Acquire("job-1", "leader-a", "worker-a", time.Second)
	if err != nil {
		t.Fatalf("acquire: %v", err)
	}
	if _, err := lm.Acquire("job-1", "leader-b", "worker-b", time.Second); err == nil {
		t.Fatal("expected active lease conflict")
	}
	if err := lm.Validate("job-1", lock.FencingToken-1); err == nil {
		t.Fatal("expected stale fencing token rejection")
	}
	now = now.Add(2 * time.Second)
	if _, err := lm.Acquire("job-1", "leader-b", "worker-b", time.Second); err != nil {
		t.Fatalf("expired lease should be recoverable: %v", err)
	}
}

func TestDAGDependenciesRetryBackoffAndCancellation(t *testing.T) {
	now := fixedTime()
	s := NewScheduler("node-a", Clock{Now: func() time.Time { return now }})
	s.BecomeLeader([]string{"node-a"}, time.Minute)

	parent, _ := s.Submit(context.Background(), JobRequest{Name: "parent", Priority: PriorityNormal, MaxAttempts: 1})
	child, _ := s.Submit(context.Background(), JobRequest{Name: "child", Priority: PriorityHigh, Dependencies: []string{parent.ID}, MaxAttempts: 1})

	job, lock, err := s.DispatchNext(context.Background(), "worker-1", time.Second)
	if err != nil {
		t.Fatalf("dispatch parent: %v", err)
	}
	if job.ID != parent.ID {
		t.Fatalf("dispatched %s, want parent %s", job.ID, parent.ID)
	}
	if err := s.Complete(context.Background(), parent.ID, lock.FencingToken, ResultCompleted, ""); err != nil {
		t.Fatalf("complete parent: %v", err)
	}
	job, _, err = s.DispatchNext(context.Background(), "worker-1", time.Second)
	if err != nil {
		t.Fatalf("dispatch child after parent completed: %v", err)
	}
	if job.ID != child.ID {
		t.Fatalf("dispatched %s, want child %s", job.ID, child.ID)
	}

	retrying, _ := s.Submit(context.Background(), JobRequest{Name: "retry", Priority: PriorityNormal, MaxAttempts: 3, InitialBackoff: time.Second})
	lockJob, retryLock, err := s.DispatchJob(context.Background(), retrying.ID, "worker-2", time.Second)
	if err != nil || lockJob.ID != retrying.ID {
		t.Fatalf("dispatch retry job: %v", err)
	}
	if err := s.Complete(context.Background(), retrying.ID, retryLock.FencingToken, ResultFailed, "temporary"); err != nil {
		t.Fatalf("complete failed retry: %v", err)
	}
	status, _ := s.GetJob(retrying.ID)
	if status.Status != StatusPending || !status.DueAt.Equal(now.Add(time.Second)) {
		t.Fatalf("retry status/due = %s/%s, want pending/%s", status.Status, status.DueAt, now.Add(time.Second))
	}

	toCancel, _ := s.Submit(context.Background(), JobRequest{Name: "cancel-me", Priority: PriorityNormal})
	if err := s.Cancel(context.Background(), toCancel.ID, "client request"); err != nil {
		t.Fatalf("cancel: %v", err)
	}
	cancelled, _ := s.GetJob(toCancel.ID)
	if cancelled.Status != StatusCancelled {
		t.Fatalf("cancelled status = %s, want %s", cancelled.Status, StatusCancelled)
	}
}

func TestHealthReportsLeaderQueuesAndRunningJobs(t *testing.T) {
	s := NewScheduler("node-a", Clock{Now: fixedTime})
	s.BecomeLeader([]string{"node-a"}, time.Minute)
	_, _ = s.Submit(context.Background(), JobRequest{Name: "queued", Priority: PriorityNormal})
	_, _, _ = s.DispatchNext(context.Background(), "worker-1", time.Second)

	health := s.Health()
	if health.NodeID != "node-a" || health.LeaderID != "node-a" || health.RunningJobs != 1 {
		t.Fatalf("unexpected health: %+v", health)
	}
}

func fixedTime() time.Time {
	return time.Date(2026, 6, 17, 12, 0, 0, 0, time.UTC)
}
